src/test/java/teammates/storage/api/NotificationsDbTest.java

```diff
@@ -0,0 +1,251 @@
+    @Test
+    public void testUpdateNotification() throws Exception {
+        final String TITLE = "title";
+        final String TITLE_DIFF = "title_diff";
+        final String MESSAGE = "message";
+        final String MESSAGE_DIFF = "message_diff";
+        final NotificationType TYPE = NotificationType.DEPRECATION;
+        final NotificationType TYPE_DIFF = NotificationType.MAINTENANCE;
+        final NotificationTargetUser TARGET_USER = NotificationTargetUser.STUDENT;
+        final NotificationTargetUser TARGET_USER_DIFF = NotificationTargetUser.GENERAL;
+        final Instant START_TIME = Instant.now();
+        final Instant START_TIME_DIFF = START_TIME.plusSeconds(600);
+        final Instant END_TIME = START_TIME.plusSeconds(3600);
+        final Instant END_TIME_DIFF = END_TIME.plusSeconds(600);
+
+        NotificationAttributes n = notificationsDb.createEntity(NotificationAttributes.builder(UUID.randomUUID().toString())
+                .withTitle(TITLE)
+                .withMessage(MESSAGE)
+                .withType(TYPE)
+                .withTargetUser(TARGET_USER)
+                .withStartTime(START_TIME)
+                .withEndTime(END_TIME)
+                .build());
+
+        ______TS("success: single field - title");
+        assertEquals(TITLE, notificationsDb.getNotification(n.getNotificationId()).getTitle());
+        assertEquals(TITLE_DIFF, notificationsDb.updateNotification(
+                NotificationAttributes.updateOptionsBuilder(n.getNotificationId())
+                        .withTitle(TITLE_DIFF)
+                        .build()).getTitle());
+        assertEquals(TITLE_DIFF, notificationsDb.getNotification(n.getNotificationId()).getTitle());
+
+        ______TS("success: single field - message");
+        assertEquals(MESSAGE, notificationsDb.getNotification(n.getNotificationId()).getMessage());
+        assertEquals(MESSAGE_DIFF, notificationsDb.updateNotification(
+                NotificationAttributes.updateOptionsBuilder(n.getNotificationId())
+                        .withMessage(MESSAGE_DIFF)
+                        .build()).getMessage());
+        assertEquals(MESSAGE_DIFF, notificationsDb.getNotification(n.getNotificationId()).getMessage());
+
+        ______TS("success: single field - type");
+        assertEquals(TYPE, notificationsDb.getNotification(n.getNotificationId()).getType());
+        assertEquals(TYPE_DIFF, notificationsDb.updateNotification(
+                NotificationAttributes.updateOptionsBuilder(n.getNotificationId())
+                        .withType(TYPE_DIFF)
+                        .build()).getType());
+        assertEquals(TYPE_DIFF, notificationsDb.getNotification(n.getNotificationId()).getType());
+
+        ______TS("success: single field - targetUser");
+        assertEquals(TARGET_USER, notificationsDb.getNotification(n.getNotificationId()).getTargetUser());
+        assertEquals(TARGET_USER_DIFF, notificationsDb.updateNotification(
```

Comment:
The main function of `testUpdateNotification` is not well tested. We should test the expected behaviour of `updateNotification`, which is to be able to :
* Retrieve the old notification, throw `EntityDoesNotExistException` if it does not exists, and 
* If it exists, call `update` in `NotificationAttributes`
* Throw `InvalidParametersException` if the notification attributes are invalid
* And of course a standard success case

For the single field tests, you may want to specify them under a separate test case e.g. `testUpdateNotification_singleFieldUpdate_shouldUpdateCorrectly` and also verify that if there's no change to the notification, it should not issue save request, so that we don't make unnecessary writes to DB.

Reply:
Thank you for the comment. I will try to improve this in the next commit.